### ISSUE TITLE
Enchancement L003: Provide an option to disallow hanging indents

### ISSUE THREAD

**Issue by dmohns** [sqlfluff/sqlfluff#3032]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Apologies if this has been discussed in the past. I have looked through old issue and Slack but couldn't quite find it. I have seen a lot of work going into L003 currently, so maybe something like this is even planned already.

I would like to start a discussion about a configuration option in SQLFluff that will allow users to disallow the usage of hanging indents as part of their linting and SQL style guides.

**Why should it be possible to configure in SQLFluff?**

_Note:_ This section is not about hanging indents specifically, but about allowing multiple indentation styles simultaneously.

My view of SQLFluff is that SQLFluff is _configurable_ linter rather than an optionated code formatter. As such, it should support users in implementing whatever rule set they desire.

Users usually choose to implement a style guide to remove ambigouity from their SQL formatting. Allowing regular indents as well as hanging indents usually achieves the opposite.

Not only are users themselves able to write extremely inconsistent SQL without SQLFluff complaining, but also the `fix` option behaves very unintuitive

For example:

```sql
WHERE TRUE
              OR FALSE
```

gets linted to

```sql
WHERE TRUE
      OR FALSE
```

while

```sql
WHERE TRUE
OR FALSE
```

gets linted to

```sql
WHERE TRUE
    OR FALSE
```

As a user that doesn't understand all the internals of SQLFluff (or any linter in general) I would expect that linting for whitespaces should have a well-defined solution.

This is because SQLFluff needs to make some decision about which style it "thinks" is correct. **A fundamental problem that cannot be solved as long as multiple options are valid.**

**Why disallow hanging indents**

Common style guides I found, don't seem to allow hanging indents.
These currently cannot be fully implemented in SQLFluff

- [dbt Style guide](https://github.com/dbt-labs/corp/blob/main/dbt_style_guide.md)
- [Mozilla Style guide](https://docs.telemetry.mozilla.org/concepts/sql_style.html#parentheses)

**Suggestion and options considered**

_Option 1:_ 

Probably in some distant future, it would be great to have a full configuration option like so

```cfg
[sqlfluff:indentation]
indentation_style = ("consistent", "hanging", "indent_unit")
```

With `consistent` being the how SQLFluff currently works. And the other two each allow only one style of indentation. (Probably there is a better word for the style that disallows hanging indents).

_Option 2:_ 

In the short term a boolean option to L003 would suffice to cover a lot of the cases mentioned above already

```cfg
[sqlfluff:indentation]
disallow_hanging_indents = True
```

**Further thoughts**

I would assume that the linting (and fixing) algorithms are also quite complicated due to the nature of allowing mixed indentation style.
This is a part not yet covered, looking for comments here.

### Dialect

all

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by WittierDinosaur** [comment]
I don't work on the linter myself - but I strongly agree this'd be a great feature!

**Comment by tunetheweb** [comment]
I think this is a good enhancement. It's been discussed a few times:
https://github.com/sqlfluff/sqlfluff/issues/365
https://github.com/sqlfluff/sqlfluff/issues/1839
https://github.com/sqlfluff/sqlfluff/issues/2853

But I think this proposal is the clearest in the ask.

I'm not sure `consistent` is the right work for the current behaviour as it will use a mixture of hanging and non-hanging, so not exactly consistent. Maybe options should be `("hanging", "indent_unit", "either")` with `either` being the current behaviour and the default?

Also this rule is one of our most complex (if not the most complex!) but @OTooleMichael cleaned it up a lot recently. Not sure if that means this will be easier to implement now, or if it's still a big task.

**Comment by OTooleMichael** [comment]
I think it would just be hanging or not. Hanging is only met in exceptional circumstances, but it also inserts itsself sometimes where it shouldn't be

```sql
CASE WHEN MAX(
                               col) THEN 1
    ELSE 2
END AS hanging```
is what hanging is doing

Apart from that indenting is fixed. Hanging often creates an indent that looks like it is pushing the indent count up by one, because its actually aligning with some INDENT (`WHERE [INDENT] TRUE`)
Variation between indent on things like JOIN and WHERE are handled else where

now Im 90% about the above, but @alanmcruickshank  should maybe weight in.
Regardless can super easily implement a no hanging option.

**Comment by dmohns** [comment]
> I'm not sure `consistent` is the right work for the current behaviour

Yeah, I meant `consistent_with_previous_lines`, which is the current rule description. But the exact wording can be discussed in case this get implemented.

> now Im 90% about the above, but @alanmcruickshank  should maybe weight in.
Regardless can super easily implement a no hanging option.

Great to hear, I think going with Option 2 (config option to disallow hanging indents) would already cover most of desired use-cases and provide a lot of value to users.

I had a quick look at L003's code (it is indeed really complex 🙈 )... Would implementing this be as easy as not calling `_handle_hanging_indents()` and `desired_indent = " " * prev_line.hanging_indent` ?

**Comment by OTooleMichael** [comment]
I think it would be that easy :)

**Comment by tunetheweb** [comment]
Closed with #3063
